Make s.add_marker accept a list of markers

magnunor opened the pull request:
This also includes several optimizations which greatly increases the speed of adding many markers at the same time: mostly avoiding calling `s.metadata`.

- [x] Add example in docstring
- [x] Add example in user guide

Commit:
s.add_marker: enable adding a list of markers + optimize for this

Lots of minor changes to avoid calling metadata for each marker

Commit:
Add test for adding markers as list

Commit:
add_marker: only update the figure after adding the last marker

Greatly increases the speed of adding many markers at the same
time with a list

Commit:
Add docstring and user guide example: adding markers using a list

magnunor commented:
Adding 1000 point markers using `s.add_marker` takes about 1 second now

francisco-dlp commented:
Great!! Maybe it'll be good to point out the huge performance advantage of adding the markers in this way in the docs.

Beside, could it take a (more general) iterable instead of a list?

magnunor commented:
Any iterable should work fine. I'll change the docstring, and add some more about this in the user guide.

Commit:
Improve docstring and user guide on adding markers as list

magnunor commented:
I changed the docstring and user guide a little bit

francisco-dlp commented on doc/user_guide/visualisation.rst:
```diff
@@ -955,8 +955,8 @@ This can be extended to 4 (or more) navigation dimensions:
 
-Many markers can added by using a for loop. Note, adding many markers
-might lead to very slow plotting.
+Many markers can added by using a for loop. Note, adding many (50+) markers
```
What about simply removing this from the User Guide: users in a hurry may miss the next paragraph which should be the preferred method in most cases.

magnunor replied:
Fixed

francisco-dlp commented on doc/user_guide/visualisation.rst:
```diff
@@ -965,16 +965,16 @@ might lead to very slow plotting.
 
     >>> from numpy.random import random
     >>> s = hs.signals.Signal2D(np.arange(300).reshape(3, 10, 10))
     >>> marker_list = []
```
What about using a generator instead?:

```python

markers = (hs.markers.point(random()*10, random()*10, size=30) for i in range(500))

```

magnunor replied:
Hmmm, doesn't work. Since I use the last marker to update the plot. `TypeError: 'generator' object is not subscriptable`. I'll try to avoid this by calling the `_draw_animated` function at the end of the `add_marker` function instead.

magnunor replied:
Fixed, and I added another example for adding different types of markers at the same time

vidartf commented on hyperspy/signal.py:
```diff
@@ -4111,45 +4114,71 @@ def add_marker(
-                    self.plot()
-                self._plot.navigator_plot.add_marker(marker)
-            marker.plot()
+        if issubclass(marker.__class__, MarkerBase):
```
Are markers iterable? If not, I would change this test to check for iterables instead (`from hyperspy.misc.utils import isiterable`). Then we can still ducktype markers without explicitly having to inherit `MarkerBase`.

magnunor replied:
Fixed! But I'm using the `issubclass(marker.__class__, MarkerBase)` when plotting the the permanent markers, since other stuff can (I guess) fairly easily sneak into the metadata.Markers.

vidartf replied:
I'm confused about where/when/why things can sneak into metadata, and not entirely sure who's responsibility it is to deal with it (ZoP: "Errors should never pass silently.").

magnunor replied:
I removed the `issubclass` `MarkerBase` check, removed the `MarkerBase` import and did some minor improvements to the `s._plot_permanent_markers` function.

Commit:
Change s.add_marker so it works with generators

Commit:
Make plotting of permanent markers more robust

Commit:
Improve user guide for adding markers as iterables

Commit:
s.add_marker: use isiterable instead of issubclass

magnunor commented:
I also added some more robustness with the updating of the plot. Now both the navigator and signal plot will be updated if they exist.

magnunor commented:
This should be ready to be merged if people are happy with it.

francisco-dlp commented:
@vidartf, are you happy with it?

vidartf commented:
What's the deal with CI on Mac?

francisco-dlp commented:
Is the question: why are some plotting tests skipped in Mac? I imagine that it has something to do with different fonts causing test failures. @ericpre should know. This was introduced in #1253 which I merged somehow prematurely thinking that the sooner it gets merged the less plotting bugs will be introduced inadvertently.

ericpre commented:
Travis does not support emulating on os x. If I remember correctly, even if we run pytest without the mpl plugin on travis os x, the tests will return an error. That's why, they are skipped on os x.
I think, this is mentioned in the dev_guide of the #1461.

francisco-dlp commented:
Why do the tests need screen emulation at all?

ericpre commented:
It looks to me, that to have the attribute `_background`, you need to create a figure, which is not possible on os x. I think this [function](https://github.com/hyperspy/hyperspy/blob/RELEASE_next_minor/hyperspy/drawing/utils.py#L103) will not work on travis os x.

Commit:
Improve _plot_permanent_markers, no longer import MarkerBase

The markers will no longer be checked to see if they inherit the
MarkerBase class

francisco-dlp commented:
@ericpre, I have the impression that matplotlib itself does not use screen simulation for their tests, so I was just wondering if a different approach is possible. In any case, maybe we can continue the discussion in #1461?

ericpre commented:
yes!

magnunor commented:
This pull request should be good to merge, unless anyone has any additional feedback

francisco-dlp commented:
@vidartf, you're the reviewer.

vidartf commented:
I'm still confused about the CI: You say that it is skipped, but for me it says 27 Failures, and it is causing Travis to fail. The question for me is then: Are the skips marked wrong? Should any of the tests of this PR be skipped for Mac? If so, that needs to be corrected before this can be merged.

vidartf commented:
Also, I would mark them as expected failures, so that we can pick up on it in the future if this suddenly starts working.

magnunor commented:
All the MacOS tests fail with the same error:

```Python
self = <hyperspy.drawing.signal1d.Signal1DFigure object at 0x13bcbe780>
    def _draw_animated(self):
        if self.ax.figure:
            canvas = self.ax.figure.canvas
>           canvas.restore_region(self._background)
E           AttributeError: 'Signal1DFigure' object has no attribute '_background'

hyperspy/drawing/figure.py:28: AttributeError
```

So I'm guessing its related to that the MacOS environment doesn't create the canvas when its supposed to (i.e. when calling `s.plot()`)?

magnunor commented:
Oooh, I guess this is due to me not marking the tests with skip for darwin

Commit:
Skip darwin pytest for test calling s.add_marker(plot_marker=True)

Also add test with s.add_marker(plot_marker=False), which should pass
on darwin

magnunor commented:
I added skips for all the tests which involve calling s.add_marker(plot_marker=True), I guess this is less then ideal but currently all the other tests involving plotting skips darwin as well

ericpre commented:
Yes, this is to expect and to be more precise, there are two different things here:
- the plotting tests are skipped on os x and this is working fine. Maybe this can be improved later.
- the tests introduce in this PR seems to call the `create_figure` or a similar function, which as it not now (and as it was before) can not work on travis os x!
However, could it be possible to run these tests without calling this function?

Commit:
Add import pytest to test_eds_tem.py

vidartf commented on hyperspy/tests/io/test_hdf5.py:
```diff
@@ -371,6 +371,7 @@ def teardown_method(self, method):
 
 class Test_permanent_markers_io:
 
+    @pytest.mark.skipif("sys.platform == 'darwin'")
```
I would prefer xfail instead (http://doc.pytest.org/en/latest/skipping.html#mark-a-test-function-as-expected-to-fail), which would allow mac users to run full test suite locally, and would also notify us if Travis setup changes..

magnunor replied:
I guess this should be done for all the current plotting tests? If so, I think that belongs in a separate pull request.

ericpre replied:
yeah, I will it in #1461.

magnunor commented:
The `s.add_marker` function relies on the create_figure function in drawing/signal1d.Signal1DFigure https://github.com/magnunor/hyperspy/blob/443ac12098d57ebc10966df14eab4fc5a34f418b/hyperspy/drawing/signal1d.py#L58 

> the tests introduce in this PR seems to call the create_figure or a similar function, which as it not now (and as it was before) can not work on os x!

Does this mean `create_figure` in `Signal1DFigure` doesn't work at all in macOS, or just the tests?

ericpre commented:
@magnunor: just the tests on travis osx.

magnunor commented:
Test are passing now that I skip the ones which require plotting in the macOS testing.

With regards to changing to xfail instead of skipif: since this is also something which should be done for all the currently "skipif" tests, this should be done in a different pull request? For example https://github.com/hyperspy/hyperspy/pull/1461 ?

vidartf commented:
Leaving it to another PR is ok, for me :)
